### PR TITLE
bugfix #18600 by using the MarkerStyle copy constructor

### PR DESCRIPTION
**Pull request by evanberkowitz**
#Deeper marker copies in Line2D update_from

Markers already support a copy construction: `MarkerStyle.set_marker`
accepts other instances of `MarkerStyle`.  However, Line2D, when
copying markers, instead construct the copy from the other instance's
`get_marker` and `get_fillstyle` methods.  This destroys (unsupported)
manipulations the user might make to the marker, as discovered in
issue #18600.

By switching constructors, we can get a more faithful copy without
explicitly relying on any of `MarkerStyle`'s private methods in places
they do not belong.

## PR Summary

This change is also more in line with the `self = other` style of the rest of the method's body.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
    - Some tests failed, but they also failed when I cloned `master`:
        - test_pcolormesh_alpha[pdf]
        - test_transparent_markers
        - test_xelatex[pdf]
        - test_pdflatex[pdf]
        - test_rcupdate
        - test_mixedmode[pdf]
        - 6 failed, 7729 passed, 89 skipped, 15 xfailed in 1031.87s (0:17:11)
    
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [X] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Review by dopplershift — APPROVED** [review]
Change seems good. Would you be able to update the docstrings for `MarkerStyle.__init__` and `MarkerStyle.set_marker()`? I had to go spelunking in the code to know that this was even correct.

**Comment by evanberkowitz** [comment]
> I had to go spelunking in the code to know that this was even correct.

Believe you me I understand THAT feeling!

I will update the doc strings (maybe tomorrow, maybe Thursday) and re-ping you here.

**Comment by evanberkowitz** [comment]
@dopplershift I updated the docs of `set_marker` and `__init__` and decided that I disagreed with the remaining flake8 criticisms.  I will 'fix' them if you require it, but I think it's extremely legible as is and that obeying the line length requirement would actually reduce legibility in this case.

File: lib/matplotlib/markers.py [hunk 1]
```diff
@@ -216,9 +217,13 @@ def __init__(self, marker=None, fillstyle=None):
         """
         Parameters
         ----------
-        marker : str or array-like or None, default: None
-            *None* means no marker. For other possible marker values see the
-            module docstring `matplotlib.markers`.
+        marker : str, array-like, Path, MarkerStyle, or None, default: None
+
```
Comment (QuLogic) [inline lib/matplotlib/markers.py:221]: No need for this first blank line in the description.

File: lib/matplotlib/markers.py [hunk 2]
```diff
@@ -283,9 +288,13 @@ def set_marker(self, marker):
 
         Parameters
         ----------
-        marker : str or array-like or None, default: None
-            *None* means no marker. For other possible marker values see the
-            module docstring `matplotlib.markers`.
+        marker : str, array-like, Path, MarkerStyle, or None, default: None
+
```
Comment (QuLogic) [inline lib/matplotlib/markers.py:292]: Ditto.

**Review by QuLogic — COMMENTED** [review]
You still have some flake8 issues as well.

**Comment by tacaswell** [comment]
Could you also add a (smoke) test that this does not blow up?

:+1: modulo fixing the style issues and adding a test.

File: lib/matplotlib/markers.py (hunk 1 above)
Comment (evanberkowitz) [reply lib/matplotlib/markers.py:221]: I wanted to provide a list of explanations for the possible `marker` values.  Do you suggest I use a
```rst
 - bullet list?
```

**Comment by evanberkowitz** [comment]
@tacaswell Can you explain what a smoke test means?  You are talking to a programming scientist, not a programmer :)

What should I be testing?  That Line2D, when given an unusual marker, actually copies the marker?  The previous implementation would have worked for all the "standard" markers.

**Comment by tacaswell** [comment]
Sorry for the jargon, it is from "turn it on and see if it smokes!".

As your original bug generated a stack trace when it should not have so the minimal test is if you can run that code and not raise (e.g. does not start smoking because it blew up).

A better and more complete test is verifying that it also actually set the marker to what it should be.

File: lib/matplotlib/markers.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/markers.py:221]: If you want a list, then yes, use a bullet list.

**Comment by evanberkowitz** [comment]
@tacaswell If I add the test, aren't we saying that the weird thing I did is in a nebulous not-supported-but-not-NOT-supported thing?

**Comment by evanberkowitz** [comment]
OK.  I actually was able to reduce my problem even further:

```
#!/usr/bin/env python3

import matplotlib
import matplotlib.pyplot as plt
from matplotlib.markers import MarkerStyle

fix, ax = plt.subplots()
plt.errorbar([1,2,3],[3,4,5], yerr=[0.5,0.5,0.5], marker=MarkerStyle('d'), linestyle='none', color='green')
plt.show()
```

which has nothing to do with my transform.  In 3.2.2 this throws a `TypeError` for either `errorbar` or `plot`, surprisingly, though `scatter` is OK.  I added a test `lib/matplotlib/tests/test_axes.py::test_marker_as_MarkerStyle` to make sure that this functionality is and remains supported.  This test does not test that the marker is correctly copied, which is what the actual code part of this PR does.

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -1317,6 +1317,14 @@ def test_arc_ellipse():
     ax.add_patch(e2)
 
 
+def test_marker_as_MarkerStyle():
+    fix, ax = plt.subplots()
+    m = mmarkers.MarkerStyle('o')
+    plt.plot([1, 2, 3], [3, 2, 1], marker=m)
+    plt.scatter([1, 2, 3], [4, 3, 2], marker=m)
+    plt.errorbar([1, 2, 3], [5, 4, 3], marker=m)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py:1325]: Use `ax`, not `plt` here.

**Review by QuLogic — COMMENTED** [review]
Still have flake8 issues. You can find them on the Files changed tab.

**Comment by tacaswell** [comment]
Modulo the `plt` vs `ax` change, the test looks conceptually right!

What you did is definitely in the "supported" category of things.

**Comment by evanberkowitz** [comment]
OK, Here's another very strange thing, but it may warrant its own, separate issue.  I suspected that my problem with my explicit legend example was with Line2D.  So I wrote this little example

```python3
#!/usr/bin/env python3

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.lines as mlines
from matplotlib.markers import MarkerStyle

print(f"{matplotlib.__version__=}")
print(f"{matplotlib.get_backend()=}")

# But when I use it in errorbar
fig_ref = plt.figure()
fig_ref.add_artist(mlines.Line2D([0.25,0.75],[0.25,0.75], marker='d'))
fig_test = plt.figure()
fig_test.add_artist(mlines.Line2D([0.25,0.75],[0.25,0.75], marker=MarkerStyle('d')))
plt.show()
```

Which outputs
```
matplotlib.__version__='3.2.2'
matplotlib.get_backend()='MacOSX'
```

markerstyle-comparison

And then I wrote a little test that would check for that difference in `test_lines.py`:

```python3
@check_figures_equal()
def test_marker_as_markerstyle_comparison(fig_test, fig_ref):
    fig_ref.add_artist(mlines.Line2D([0.25,0.75],[0.25,0.75], marker='d'))
    fig_test.add_artist(mlines.Line2D([0.25,0.75],[0.25,0.75], marker=MarkerStyle('d')))
```

and the test passes!

```
> pytest lib/matplotlib/tests/test_lines.py::test_marker_as_markerstyle_comparison
=========================== test session starts ============================
platform darwin -- Python 3.8.5, pytest-6.1.0, py-1.9.0, pluggy-0.13.1
rootdir: /Users/evanberkowitz/src/matplotlib, configfile: pytest.ini
collected 3 items

lib/matplotlib/tests/test_lines.py ...                               [100%]

============================ 3 passed in 2.62s =============================
```

What?!  I tried adding to the above test
```python3
    fig_ref.savefig('ref')
    fig_test.savefig('test')
```
to the test to see if they're REALLY the same or not and sure enough, the both look like this:

test

What gives?

Note that I observe a difference in 3.2.2 but not the latest that my PR is written against.

**Comment by QuLogic** [comment]
That was fixed in #16692, I think.

**Comment by evanberkowitz** [comment]
OK, that seems correct.  I will not add my test then.

**Comment by evanberkowitz** [comment]
@QuLogic---as I mentioned in my comment towards the top of this PR, if you actually look at the code for the flake8 complaints and add a newline, the docstrings become _harder_ to read.  If you tell me that you understand that, then I will change them.  Otherwise, I advocate ignoring those complaints.

**Comment by QuLogic** [comment]
We do not ignore flake8 arbitrarily. We have plenty of these lists wrapped, so I don't see how it is less legible.

**Comment by evanberkowitz** [comment]
[Widows and orphans](https://en.wikipedia.org/wiki/Widows_and_orphans) are known to typesetters to cause ugly looking or hard-to-read text.  In this case the enormous amount of whitespace introduced by the orphans created by wrapping these only-slightly-over-length lines cause visual interruption.  A linter usually doesn't take that sort of "global appearance" into account.  But I'll change it against my better judgement.

**Comment by evanberkowitz** [comment]
Um... now eslint simply failed to set up?  I don't think this one was my fault.

**Comment by evanberkowitz** [comment]
eslint reran and succeeded, so all tests have now passed.

**Comment by tacaswell** [comment]
> But I'll change it against my better judgement.

On projects that only have a handful (or just one) person working on them it is plausible to have use your judgement on the formatting, but on a project the scale of Matplotlib (we have over 1k people who have touched the code in some way at this point!) we have to rely on automated tools to maintain any consistency in the codebase.  It also means we can have the robot tell you the style is wrong than have a person tell you ;)

**Comment by tacaswell** [comment]
Thanks for finding and fixing this @evanberkowitz.

Congratulations on your first Matplotlib PR :tada:  hopefully we will hear from you again!

**Comment by evanberkowitz** [comment]
Glad to contribute.  Hopefully there will never be further bugs for me to encounter ;) !

**Pull request merged**